File: OpenTokFs.RequestTypes/OpenTokArchiveStartRequest.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;

namespace OpenTokFs.RequestTypes {
    [JsonConverter(typeof(OpenTokArchiveStartRequestPropertyConverter))]
    public class OpenTokArchiveStartRequest {
        public string SessionId { get; }

        public OpenTokVideoLayout Layout { get; set; } = OpenTokVideoLayout.BestFit;
        public bool HasAudio { get; set; } = true;
        public bool HasVideo { get; set; } = true;
        public string? Name { get; set; } = null;
        public string OutputMode { get; set; } = "composed";
        public string Resolution { get; set; } = "640x480";

        public OpenTokArchiveStartRequest(string sessionId) {
            this.SessionId = sessionId ?? throw new ArgumentNullException(nameof(sessionId));
        }
    }

    public class OpenTokArchiveStartRequestPropertyConverter : JsonConverter {
        public override bool CanConvert(Type? objectType) {
            return objectType == typeof(OpenTokArchiveStartRequest);
        }

        public override object? ReadJson(JsonReader? reader, Type? objectType, object? existingValue, JsonSerializer? serializer) {
            return false;
        }

        public override void WriteJson(JsonWriter? writer, object? value, JsonSerializer? serializer) {
            if (!(value is OpenTokArchiveStartRequest obj)) {
                serializer!.Serialize(writer!, value);
                return;
            }

            var dict = new Dictionary<string, object?> {
                ["sessionId"] = obj.SessionId,
                ["hasAudio"] = obj.HasAudio,
                ["hasVideo"] = obj.HasVideo,
                ["name"] = obj.Name,
                ["outputMode"] = obj.OutputMode
            };
            if (obj.OutputMode == "composed") {
                dict.Add("layout", obj.Layout);
                dict.Add("resolution", obj.Resolution);
            }
            serializer!.Serialize(writer!, dict);
        }
    }
}

File: OpenTokFs/RequestTypes/OpenTokArchiveStartRequest.fs
namespace OpenTokFs.RequestTypes

open Newtonsoft.Json

[<JsonConverter(typeof<OpenTokArchiveStartRequestPropertyConverter>)>]
type OpenTokArchiveStartRequest(sessionId: string) =
    member __.sessionId = sessionId

    member val layout: OpenTokVideoLayout = OpenTokVideoLayout.BestFit with get, set
    member val hasAudio: bool = true with get, set
    member val hasVideo: bool = true with get, set
    member val name: string = null with get, set
    member val outputMode: string = "composed" with get, set
    member val resolution: string = "640x480" with get, set

and OpenTokArchiveStartRequestPropertyConverter() =
    inherit JsonConverter()

    override _.CanConvert t =
        t = typeof<OpenTokArchiveStartRequest>

    override _.ReadJson (reader, objectType, existingValue, serializer) =
        base.ReadJson (reader, objectType, existingValue, serializer)

    override _.WriteJson (writer, value, serializer) =
        match value with
        | :? OpenTokArchiveStartRequest as obj ->
            serializer.Serialize (writer, dict [
                ("sessionId", obj.sessionId :> obj)
                ("hasAudio", obj.hasAudio :> obj)
                ("hasVideo", obj.hasVideo :> obj)
                ("name", obj.name :> obj)
                ("outputMode", obj.outputMode :> obj)
                if obj.outputMode = "composed" then
                    //("layout", obj.layout :> obj)
                    ("resolution", obj.resolution :> obj)
            ])
        | _ ->
            serializer.Serialize (writer, value)
